### PR TITLE
JS: Reorganize Closure module

### PR DESCRIPTION
**Pull request by asger-semmle**
First of all, sorry for the horrendous diff in the second commit.

The class for `goog.require` calls needs to be extensible and can't be based on the AST, as it must be possible to use an `import` node here. The refactoring in commit nr 2 leads up to the commit that makes this possible.

To contribute to the confusing mess that is commit nr 2, I opted for a midway renaming of most of the classes to be more consistent. We now consistently use the term "closure namespace" instead of "library access path" and "namespace id".

Currently running evaluation on closure.slugs

### CODE REVIEW

File: javascript/ql/src/semmle/javascript/Closure.qll [hunk 1]
```diff
@@ -6,101 +6,83 @@ import javascript
...
+    /**
+     * Gets the namespace being referenced.
+     */
+    abstract string getClosureNamespace();
   }
 
   /**
-   * An expression statement consisting of a call to a function
-   * in the `goog` namespace.
+   * A dataflow node that returns the value of a closure namespace.
```
Comment (xiemaisi) [inline javascript/ql/src/semmle/javascript/Closure.qll]:
```suggestion
   * A data flow node that returns the value of a Closure namespace.
```

Comment (xiemaisi) [reply javascript/ql/src/semmle/javascript/Closure.qll]: Ditto in other places

File: javascript/ql/src/semmle/javascript/Closure.qll [hunk 2]
```diff
@@ -6,101 +6,83 @@ import javascript
...
-    /** Gets an argument to the invoked function. */
-    Expr getAnArgument() { result = getArgument(_) }
+    override string getClosureNamespace() { result = getArgument(0).asExpr().getStringValue() }
   }
 
-  abstract private class GoogNamespaceRef extends ExprOrStmt { abstract string getNamespaceId(); }
-
   /**
-   * A call to `goog.provide`.
+   * Holds if `node` is the dataflow node for a top-level expression.
```
Comment (xiemaisi) [inline javascript/ql/src/semmle/javascript/Closure.qll]: The term "top-level expression" deserves an explanation or example.

Comment (asger-semmle) [reply javascript/ql/src/semmle/javascript/Closure.qll]: I've elaborated the qldoc a bit.

Or did you mean why we bother with this predicate at all?
- that's how it worked before the PR.
- less chance of picking up random stuff in a minified file.
- more closely matches how the extractor recognizes `goog.module`.

**Comment by asger-semmle** [comment]
> I assume you've thought about the advantages and disadvantages of going with a plain abstract class-based design

`Closure::moduleImport` is the main way to get accesses to a closure namespace, so it didn't seem likely that anyone would use this class for general query work. But we've been wrong about this before, I guess. Would you like me to change it?

> Are these abstract classes ever meant to be extended outside the library?

Yes, the motivation for this was to make `Closure::moduleImport` extensible, which is now possible through `ClosureNamespaceAccess`.

File: javascript/ql/src/semmle/javascript/Closure.qll (hunk 2 above)
Comment (xiemaisi) [reply javascript/ql/src/semmle/javascript/Closure.qll]: The expanded QLDoc is what I was after, thanks.

**Comment by xiemaisi** [comment]
I think I'd prefer using the facade pattern unless you think it adds too much complexity. If we have clearly documented, distinct ways of subclassing for extension on the one hand and of subclassing for refinement on the other hand, then we'll avoid someone else (or indeed ourselves) accidentally using the wrong one, which is tricky to clean up as we saw with the `SourceNode` business.

**Comment by xiemaisi** [comment]
LGTM modulo evaluation and the refactoring discussed above.

**Comment by asger-semmle** [comment]
[Evaluation](https://git.semmle.com/gist/asger/695fb54c4c1841bc3941358a95e84d67) was horrendous.

I suspect it's because of the dependency: `ClosureModule -> ClosureNamespaceRef -> ClosureNamespaceAccess -> SourceNode -> ModuleImportNode -> Module -> ClosureModule`. I'm working on fixing that.

**Comment by esben-semmle** [comment]
FYI. You may want to look into another performance problem first:
`definitions.ql` is timing out during `TypeInference::AnalyzedNode::getALocalValue#ff#cur_delta` on master for `g/google/closure-library` when run on an azure worker.

**Comment by asger-semmle** [comment]
Cutting the dependency chain fixed the issue on [mxnet.js](https://git.semmle.com/asger/dist-compare-reports/tree/domitian.ti.semmle.com_1550235338959) in isolation. Running more evaluations with the `::Range` pattern included.

**Comment by asger-semmle** [comment]
[Evaluation](https://git.semmle.com/asger/dist-compare-reports/tree/closure-reorg) looks good now. This should be ready to merge.

**Pull request merged**